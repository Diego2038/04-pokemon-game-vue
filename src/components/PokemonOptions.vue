<template>
  <div class="options-container">
    <ul>
      <li v-for="pokemon in pokemonsArray" 
      :key="pokemon.id"
      @click="$emit( 'selection', pokemon, 'dos', 'tres' )" > <!-- Puedes enviar mÃ¡s de un argumento -->
        {{pokemon.name}}
      </li>
    </ul>
  </div>
</template>

<script>
// selection es el nombre del evento que emites, el padre lo recibe y lo consume de ser necesario
export default {
  name: 'pokemonOptions',
  props : {
    pokemonsArray : {
      type : Array,
      required : true
    }
  }
}
</script>

<style scoped>
ul {
    list-style-type: none;
}
li {
    background-color: white;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
}

li:hover {
    background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
    display: flex;
    justify-content: center;
}
</style>