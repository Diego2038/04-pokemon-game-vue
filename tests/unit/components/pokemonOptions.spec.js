import { shallowMount } from "@vue/test-utils";
import PokemonOptions from "@/components/PokemonOptions.vue";
import { pokemons } from "../mocks/pokemons.mock";

describe("PokemonOptions component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonOptions, {
      props: {
        pokemonsArray: pokemons,
      },
    });
  });

  test("Debe de hacer Match con el snapshot", () => {
    // console.log( wrapper.html() )
    expect(wrapper.html()).toMatchSnapshot();

    // Con esta no generas un archivo snapshot, pero es más sensible al cambio
    expect(wrapper.html()).toMatchInlineSnapshot(`
      <div class="options-container">
        <ul>
          <li>
            <!-- Puedes enviar más de un argumento --> bulbasaur
          </li>
          <li>
            <!-- Puedes enviar más de un argumento --> ivysaur
          </li>
          <li>
            <!-- Puedes enviar más de un argumento --> venusaur
          </li>
          <li>
            <!-- Puedes enviar más de un argumento --> charmander
          </li>
        </ul>
      </div>
    `);
  });

  test("Debe de mostrar las 4 opciones correctas", () => {
    // Que existan 4 li como opciones
    const liArrays = wrapper.findAll("li");
    expect(liArrays.length).toBe(4);

    // cada li debe de tener su nombre del pokemon
    for (let i = 0; i < liArrays.length; i++) {
      expect(liArrays[i].text()).toBe(pokemons[i].name);
    }
  });

  test('Debe emitir "selection" con sus respectivos parámetros al hacer click', () => {
    const [l1, l2, l3, l4] = wrapper.findAll("li");

    l1.trigger("click"); 
    l2.trigger("click"); 
    l3.trigger("click"); 
    l4.trigger("click"); 

    // console.log(wrapper.emitted("selection")); // Nótese que retorna un arreglo de todos los parámetros que retorna el emit
    expect( wrapper.emitted('selection').length ).toBe(4)

    expect( wrapper.emitted('selection')[0] ).toEqual( [{ pokemon: { name: 'bulbasaur', id: 1 }, dos: 'dos', tres: 'tres'}]) // OJO, agregué intencionalmente el componente 'dos' y 'tres' para "practicar" que la emisión de eventos se pueden agregar más de un parámetro (tantos como uno quiera)
    expect( wrapper.emitted('selection')[1] ).toEqual( [{ pokemon: { name: 'ivysaur', id: 2 }, dos: 'dos', tres: 'tres'}])
    expect( wrapper.emitted('selection')[2] ).toEqual( [{ pokemon: { name: 'venusaur', id: 3 }, dos: 'dos', tres: 'tres'}])
    expect( wrapper.emitted('selection')[3] ).toEqual( [{ pokemon: { name: 'charmander', id: 4 }, dos: 'dos', tres: 'tres'}])
    

  });
});
