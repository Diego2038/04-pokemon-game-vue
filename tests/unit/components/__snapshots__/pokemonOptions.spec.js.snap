// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions component Debe de hacer Match con el snapshot 1`] = `
<div class="options-container">
  <ul>
    <li>
      <!-- Puedes enviar m치s de un argumento --> bulbasaur
    </li>
    <li>
      <!-- Puedes enviar m치s de un argumento --> ivysaur
    </li>
    <li>
      <!-- Puedes enviar m치s de un argumento --> venusaur
    </li>
    <li>
      <!-- Puedes enviar m치s de un argumento --> charmander
    </li>
  </ul>
</div>
`;
